### PR TITLE
Makes grids that pass a certain size airtight.

### PR DESCRIPTION
**Pull request by moonheart08**
## About the PR 
Grids larger than seven tiles now automatically have the ability to hold air.
This is, _iirc_, Vera Approved:tm:

**Changelog**
:cl: moony
- add: Player built shuttles should now be able to hold an atmosphere, assuming they're eight tiles or larger.

### CODE REVIEW

File: Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            return;
+
+        // We can't actually count how many tiles there are efficiently, so instead estimate with the mass.
+        if (physics.Mass / ShuttleSystem.TileMassMultiplier >= 7.0f && !HasComp<IAtmosphereComponent>(uid))
+        {
+            AddComp<GridAtmosphereComponent>(uid);
+            Logger.InfoS("atmos", $"Giving grid {uid} GridAtmosphereComponent.");
+        }
+        // It's not super important to remove it should the grid become too small again.
+        // If explosions ever gain the ability to outright shatter grids, do rethink this.
```
Comment (metalgearsloth) [inline Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs:44]: With electro's explosion PR it probably can; if a grid hits 0 tiles it gets deleted.

File: Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        if (!((e.OldTile.IsSpace(_tileDefinitionManager) && !e.NewTile.IsSpace(_tileDefinitionManager)) ||
+            (!e.OldTile.IsSpace(_tileDefinitionManager) && e.NewTile.IsSpace(_tileDefinitionManager))))
+            return;
+
+        var uid = _mapManager.GetGridEuid(e.NewTile.GridIndex);
+        if (!TryComp<PhysicsComponent>(uid, out var physics))
+            return;
+
+        // We can't actually count how many tiles there are efficiently, so instead estimate with the mass.
+        if (physics.Mass / ShuttleSystem.TileMassMultiplier >= 7.0f && !HasComp<IAtmosphereComponent>(uid))
```
Comment (metalgearsloth) [inline Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs:38]: Alternatively could listen to grid fixture changes (only issues when space / non-space changes) or we make the chunk validtiles property also work on grids.

File: Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+public sealed class AutomaticAtmosSystem : EntitySystem
+{
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly ITileDefinitionManager _tileDefinitionManager = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        _mapManager.TileChanged += MapManagerOnTileChanged;
+    }
```
Comment (metalgearsloth) [inline Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs:21]: This needs to be unsubscribed in shutdown.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    private void MapManagerOnTileChanged(object? sender, TileChangedEventArgs e)
+    {
+        // Only if a atmos-holding tile has been added or removed.
+        // Also, these calls are surprisingly slow.
+        // TODO: Make tiledefmanager cache the IsSpace property, and turn this lookup-through-two-interfaces into
+        // TODO: a simple array lookup, as tile IDs are likely contiguous, and there's at most 2^16 possibilities anyway.
+        if (!((e.OldTile.IsSpace(_tileDefinitionManager) && !e.NewTile.IsSpace(_tileDefinitionManager)) ||
+            (!e.OldTile.IsSpace(_tileDefinitionManager) && e.NewTile.IsSpace(_tileDefinitionManager))))
+            return;
```
Comment (ElectroJr) [inline Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs:31]: For now, checking `Tile.IsEmpty` before calling `Tile.IsSpace()` might help speed things up a bit. Though maybe IsSpace should just do that automatically?

Comment (metalgearsloth) [reply Content.Server/Atmos/EntitySystems/AutomaticAtmosSystem.cs:31]: Ideally. When we get diagonal tiles a tile won't necessarily be empty to be space.

Realistically I doubt these calls will show up on a profiler.

**Pull request merged**